### DISCUSSION TITLE
Add or Remove MorphTarget

### DISCUSSION THREAD

**Discussion by August1996** [google/filament#5800] [Q&A]
How can i add or remove some morph targets to a primitive?

The `MorphTargetBuffer` may only supports to replace a morph target.

@daemyung @bejado @prideout

**Comment by daemyung** [comment]
@August1996 `MorphTargetBuffer` is the bulk memory containing all morph targets and the size of this isn't flexible. So if you want to add or remove then you should create the another `MorphTargetBuffer`. There is a reasons to this limitation. GPU has the limitation that textures can be bounded. Usually the number of bounding slot is less then the morph target count. To avoid this limitation, we upload morph targets to texture 2d array. Thus supporting dynamic size is impossible.

I can give you some WA to satisfy your needs. Allocate `MorphTargetBuffer` with the max morph target count then set morph target data to 0 or set weight 0 to morph targets which you want to remove.

**Reply by August1996** [reply]
@daemyung `Allocate MorphTargetBuffer with the max morph target count` may results in some performance issues or memory issues?

**Reply by daemyung** [reply]
Yes more memory is needed and not sure about performance.

**Reply by August1996** [reply]
> Yes more memory is needed and not sure about performance.

so maybe the best opportunity to add or remove morph targets is cgltf parsing process.

**Reply by August1996** [reply]
> Yes more memory is needed and not sure about performance.

The morph target calculating will stop if the weight equals zero?

**Reply by daemyung** [reply]
> The morph target calculating will stop if the weight equals zero?

No just skip morph targets.
